// itrm-tools/dist/api/HttpsAPIService.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpsAPIService = void 0;
const https_1 = __importDefault(require("https"));
const APIService_1 = require("./APIService");
class HttpsAPIService extends APIService_1.APIService {
    constructor(config, options) {
        super(config);
        this.httpsServer = undefined;
        this.options = options;
    }
    init() {
        super.init();
        this.httpsServer = https_1.default.createServer(this.options, this.app);
    }
    run(init) {
        return new Promise((resolve, reject) => {
            this.server = this.httpsServer.listen(this.port, () => __awaiter(this, void 0, void 0, function* () {
                console.log(`⚡️[server]: Server is running at https://localhost:${this.port}`);
                try {
                    yield init();
                    resolve({ status: 'Ok', message: 'Server has been Initialized' });
                }
                catch (err) {
                    reject({
                        status: 'Error',
                        message: 'An error has ocurred in the initialization function',
                        error: err
                    });
                }
            }));
        });
    }
}
exports.HttpsAPIService = HttpsAPIService;
